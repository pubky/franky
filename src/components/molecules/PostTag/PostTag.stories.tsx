import type { Meta, StoryObj } from '@storybook/nextjs-vite';

import { PostTag } from './PostTag';

const meta: Meta<typeof PostTag> = {
  title: 'Molecules/Post/PostTag',
  component: PostTag,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
      values: [
        { name: 'dark', value: '#05050A' },
        { name: 'light', value: '#F2F2F7' },
      ],
    },
    docs: {
      description: {
        component:
          'Tag component for posts featuring dynamic colors, counter, and removable state. Uses colors from utils to generate consistent tag colors.',
      },
    },
  },
  args: {
    label: 'bitcoin',
  },
  argTypes: {
    onClick: { action: 'clicked' },
    onClose: { action: 'closed' },
    color: { control: 'color' },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Playground: Story = {
  render: (args) => <PostTag {...args} />,
};

// States without counter or close
export const Default: Story = {
  args: {
    label: 'bitcoin',
    selected: false,
    showClose: false,
  },
  render: (args) => <PostTag {...args} />,
};

export const Hover: Story = {
  args: {
    label: 'bitcoin',
    selected: false,
    showClose: false,
  },
  render: (args) => <PostTag {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Hover state shows an inset glow effect matching the tag color. Try hovering over the tag.',
      },
    },
    pseudo: { hover: true },
  },
};

export const Selected: Story = {
  args: {
    label: 'bitcoin',
    selected: true,
    showClose: false,
  },
  render: (args) => <PostTag {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Selected state shows a solid border matching the tag color.',
      },
    },
  },
};

// With counter
export const WithCounter: Story = {
  args: {
    label: 'bitcoin',
    count: 16,
    selected: false,
    showClose: false,
  },
  render: (args) => <PostTag {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Tag with counter showing the number of posts.',
      },
    },
  },
};

export const WithCounterSelected: Story = {
  args: {
    label: 'bitcoin',
    count: 16,
    selected: true,
    showClose: false,
  },
  render: (args) => <PostTag {...args} />,
};

// With close button
export const WithClose: Story = {
  args: {
    label: 'bitcoin',
    selected: false,
    showClose: true,
  },
  render: (args) => <PostTag {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Tag with close button for removing the tag.',
      },
    },
  },
};

export const WithCloseSelected: Story = {
  args: {
    label: 'bitcoin',
    selected: true,
    showClose: true,
  },
  render: (args) => <PostTag {...args} />,
};

// All color variants
export const ColorVariants: Story = {
  render: () => (
    <div className="flex flex-wrap gap-2">
      <PostTag label="bitcoin" color="#FF9900" />
      <PostTag label="synonym" color="#FF0000" />
      <PostTag label="pubky" color="#FC00FF" />
      <PostTag label="nostr" color="#004BFF" />
      <PostTag label="lightning" color="#00F0FF" />
      <PostTag label="sats" color="#00FF5D" />
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Different color variants. Colors are automatically generated from tag labels using generateRandomColor utility.',
      },
    },
  },
};

// All states showcase
export const AllStates: Story = {
  render: () => (
    <div className="flex flex-col gap-4">
      <div className="flex flex-wrap gap-2">
        <PostTag label="bitcoin" />
        <PostTag label="bitcoin" showClose />
        <PostTag label="bitcoin" selected />
        <PostTag label="bitcoin" showClose selected />
      </div>
      <div className="flex flex-wrap gap-2">
        <PostTag label="bitcoin" count={16} />
        <PostTag label="bitcoin" count={16} showClose />
        <PostTag label="bitcoin" count={16} selected />
        <PostTag label="bitcoin" count={16} showClose selected />
      </div>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Showcase of all possible states: default, with close, selected, with counter.',
      },
    },
  },
};

// Interactive demo with callbacks
export const WithCallbacks: Story = {
  args: {
    label: 'bitcoin',
    count: 16,
    showClose: true,
    selected: false,
    onClick: () => console.log('Tag clicked'),
    onClose: () => console.log('Close clicked'),
  },
  render: (args) => <PostTag {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Interactive tag with click and close callbacks. Check the console when clicking.',
      },
    },
  },
};

// Real-world usage example
export const TagList: Story = {
  render: () => (
    <div className="flex max-w-md flex-wrap gap-2">
      <PostTag label="bitcoin" count={142} selected />
      <PostTag label="synonym" count={89} />
      <PostTag label="pubky" count={56} />
      <PostTag label="nostr" count={34} />
      <PostTag label="lightning" count={23} />
      <PostTag label="sats" count={12} />
      <PostTag label="hodl" count={8} />
      <PostTag label="btc" count={156} selected />
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Example of a tag list as it might appear in a real application.',
      },
    },
  },
};
