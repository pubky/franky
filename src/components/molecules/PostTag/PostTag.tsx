import * as React from 'react';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

export interface PostTagProps {
  /** Tag label text */
  label: string;
  /** Number of posts with this tag (optional) */
  count?: number;
  /** Show the close/remove button */
  showClose?: boolean;
  /** Selected state */
  selected?: boolean;
  /** Callback when tag is clicked */
  onClick?: () => void;
  /** Callback when close button is clicked */
  onClose?: (e: React.MouseEvent) => void;
  /** Custom color (hex) for the tag - if not provided, generates from label */
  color?: string;
  /** Additional className */
  className?: string;
}

export function PostTag({
  label,
  count,
  showClose = false,
  selected = false,
  onClick,
  onClose,
  color,
  className,
}: PostTagProps) {
  const tagColor = color || Libs.generateRandomColor(label);
  const backgroundGradient = `linear-gradient(90deg, ${Libs.hexToRgba('#05050A', 0.7)} 0%, ${Libs.hexToRgba('#05050A', 0.7)} 100%), linear-gradient(90deg, ${tagColor} 0%, ${tagColor} 100%)`;

  const handleClose = (e: React.MouseEvent) => {
    e.stopPropagation();
    onClose?.(e);
  };

  return (
    <Atoms.Toggle
      pressed={selected}
      onClick={onClick}
      className={Libs.cn(
        'group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg',
        'border-0 text-sm font-bold leading-5 text-white',
        'transition-all duration-200',
        // Override Toggle default hover styles - keep text white
        'hover:bg-transparent hover:text-white',
        // Selected state - add border
        'data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white',
        className,
      )}
      style={{
        backgroundImage: backgroundGradient,
        boxShadow: selected ? `inset 0 0 0 1px ${tagColor}` : undefined,
      }}
      aria-label={count !== undefined ? `${label} tag (${count} posts)` : `${label} tag`}
    >
      {/* Tag content */}
      <span className="flex items-center gap-1.5 text-sm leading-5">
        <span className="font-bold">{label}</span>
        {count !== undefined && <span className="font-medium opacity-50">{count}</span>}
      </span>

      {/* Close button */}
      {showClose && (
        <span
          onClick={handleClose}
          className="flex size-4 shrink-0 cursor-pointer items-center justify-center rounded opacity-70 transition-opacity hover:opacity-100"
          aria-label={`Remove ${label} tag`}
          role="button"
        >
          <Libs.X className="size-3" strokeWidth={2} />
        </span>
      )}

      {/* Hover shadow overlay - exactly as Figma */}
      <div
        className="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
        style={{
          boxShadow: `inset 0px 0px 8px 0px ${tagColor}`,
        }}
        aria-hidden="true"
      />

      {/* Selected border overlay */}
      {selected && (
        <div
          className="pointer-events-none absolute inset-0 rounded-lg border border-solid"
          style={{ borderColor: tagColor }}
          aria-hidden="true"
        />
      )}
    </Atoms.Toggle>
  );
}
