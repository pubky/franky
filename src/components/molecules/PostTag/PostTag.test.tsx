import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { PostTag } from './PostTag';

// Mock @/libs with partial mock
vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
    generateRandomColor: vi.fn((str: string) => {
      // Return consistent colors for testing
      const colorMap: Record<string, string> = {
        bitcoin: '#FF9900',
        synonym: '#FF0000',
        test: '#00FF00',
      };
      return colorMap[str.toLowerCase()] || '#FF9900';
    }),
    hexToRgba: vi.fn((hex: string, alpha: number) => {
      const [r, g, b] = hex.match(/\w\w/g)!.map((x) => parseInt(x, 16));
      return `rgba(${r}, ${g}, ${b}, ${alpha})`;
    }),
    X: ({ className, strokeWidth }: { className?: string; strokeWidth?: number }) => (
      <div data-testid="x-icon" className={className} data-stroke-width={strokeWidth}>
        X
      </div>
    ),
  };
});

describe('PostTag', () => {
  it('calls onClick when tag is clicked', () => {
    const mockOnClick = vi.fn();
    render(<PostTag label="bitcoin" onClick={mockOnClick} />);

    const tag = screen.getByRole('button');
    fireEvent.click(tag);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('calls onClose when close button is clicked', () => {
    const mockOnClose = vi.fn();
    render(<PostTag label="bitcoin" showClose onClose={mockOnClose} />);

    const closeButton = screen.getByLabelText(/remove bitcoin tag/i);
    fireEvent.click(closeButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('does not call onClick when close button is clicked', () => {
    const mockOnClick = vi.fn();
    const mockOnClose = vi.fn();
    render(<PostTag label="bitcoin" showClose onClick={mockOnClick} onClose={mockOnClose} />);

    const closeButton = screen.getByLabelText(/remove bitcoin tag/i);
    fireEvent.click(closeButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
    expect(mockOnClick).not.toHaveBeenCalled();
  });

  it('updates aria-label when count is provided', () => {
    const { rerender } = render(<PostTag label="bitcoin" />);
    expect(screen.getByLabelText(/bitcoin tag$/i)).toBeInTheDocument();

    rerender(<PostTag label="bitcoin" count={16} />);
    expect(screen.getByLabelText(/bitcoin tag \(16 posts\)/i)).toBeInTheDocument();
  });

  it('renders with custom color', () => {
    const { container } = render(<PostTag label="bitcoin" color="#123456" />);
    const tag = screen.getByRole('button');

    // Verify the tag is rendered (custom color is applied via inline styles in the snapshot)
    expect(tag).toBeInTheDocument();
    expect(container.firstChild).toBeTruthy();
  });
});

describe('PostTag - Snapshots', () => {
  it('matches snapshot with default state', () => {
    const { container } = render(<PostTag label="bitcoin" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot when selected', () => {
    const { container } = render(<PostTag label="bitcoin" selected />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with counter', () => {
    const { container } = render(<PostTag label="bitcoin" count={16} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with close button', () => {
    const { container } = render(<PostTag label="bitcoin" showClose />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with all features', () => {
    const { container } = render(<PostTag label="bitcoin" count={16} showClose selected />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with custom color', () => {
    const { container } = render(<PostTag label="bitcoin" color="#FF0000" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
