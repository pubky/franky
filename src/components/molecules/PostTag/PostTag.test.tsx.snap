// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostTag - Snapshots > matches snapshot when selected 1`] = `
<button
  aria-label="bitcoin tag"
  aria-pressed="true"
  class="cursor-pointer inline-flex items-center justify-center bg-transparent focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 min-w-9 group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg border-0 text-sm font-bold leading-5 text-white transition-all duration-200 hover:bg-transparent hover:text-white data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white"
  data-state="on"
  style="box-shadow: inset 0 0 0 1px #FF9900;"
  type="button"
>
  <span
    class="flex items-center gap-1.5 text-sm leading-5"
  >
    <span
      class="font-bold"
    >
      bitcoin
    </span>
  </span>
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
    style="box-shadow: inset 0px 0px 8px 0px #FF9900;"
  />
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg border border-solid"
    style="border-color: rgb(255, 153, 0);"
  />
</button>
`;

exports[`PostTag - Snapshots > matches snapshot with all features 1`] = `
<button
  aria-label="bitcoin tag (16 posts)"
  aria-pressed="true"
  class="cursor-pointer inline-flex items-center justify-center bg-transparent focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 min-w-9 group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg border-0 text-sm font-bold leading-5 text-white transition-all duration-200 hover:bg-transparent hover:text-white data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white"
  data-state="on"
  style="box-shadow: inset 0 0 0 1px #FF9900;"
  type="button"
>
  <span
    class="flex items-center gap-1.5 text-sm leading-5"
  >
    <span
      class="font-bold"
    >
      bitcoin
    </span>
    <span
      class="font-medium opacity-50"
    >
      16
    </span>
  </span>
  <span
    aria-label="Remove bitcoin tag"
    class="flex size-4 shrink-0 cursor-pointer items-center justify-center rounded opacity-70 transition-opacity hover:opacity-100"
    role="button"
  >
    <div
      class="size-3"
      data-stroke-width="2"
      data-testid="x-icon"
    >
      X
    </div>
  </span>
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
    style="box-shadow: inset 0px 0px 8px 0px #FF9900;"
  />
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg border border-solid"
    style="border-color: rgb(255, 153, 0);"
  />
</button>
`;

exports[`PostTag - Snapshots > matches snapshot with close button 1`] = `
<button
  aria-label="bitcoin tag"
  aria-pressed="false"
  class="cursor-pointer inline-flex items-center justify-center bg-transparent focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 min-w-9 group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg border-0 text-sm font-bold leading-5 text-white transition-all duration-200 hover:bg-transparent hover:text-white data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white"
  data-state="off"
  type="button"
>
  <span
    class="flex items-center gap-1.5 text-sm leading-5"
  >
    <span
      class="font-bold"
    >
      bitcoin
    </span>
  </span>
  <span
    aria-label="Remove bitcoin tag"
    class="flex size-4 shrink-0 cursor-pointer items-center justify-center rounded opacity-70 transition-opacity hover:opacity-100"
    role="button"
  >
    <div
      class="size-3"
      data-stroke-width="2"
      data-testid="x-icon"
    >
      X
    </div>
  </span>
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
    style="box-shadow: inset 0px 0px 8px 0px #FF9900;"
  />
</button>
`;

exports[`PostTag - Snapshots > matches snapshot with counter 1`] = `
<button
  aria-label="bitcoin tag (16 posts)"
  aria-pressed="false"
  class="cursor-pointer inline-flex items-center justify-center bg-transparent focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 min-w-9 group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg border-0 text-sm font-bold leading-5 text-white transition-all duration-200 hover:bg-transparent hover:text-white data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white"
  data-state="off"
  type="button"
>
  <span
    class="flex items-center gap-1.5 text-sm leading-5"
  >
    <span
      class="font-bold"
    >
      bitcoin
    </span>
    <span
      class="font-medium opacity-50"
    >
      16
    </span>
  </span>
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
    style="box-shadow: inset 0px 0px 8px 0px #FF9900;"
  />
</button>
`;

exports[`PostTag - Snapshots > matches snapshot with custom color 1`] = `
<button
  aria-label="bitcoin tag"
  aria-pressed="false"
  class="cursor-pointer inline-flex items-center justify-center bg-transparent focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 min-w-9 group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg border-0 text-sm font-bold leading-5 text-white transition-all duration-200 hover:bg-transparent hover:text-white data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white"
  data-state="off"
  type="button"
>
  <span
    class="flex items-center gap-1.5 text-sm leading-5"
  >
    <span
      class="font-bold"
    >
      bitcoin
    </span>
  </span>
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
    style="box-shadow: inset 0px 0px 8px 0px #FF0000;"
  />
</button>
`;

exports[`PostTag - Snapshots > matches snapshot with default state 1`] = `
<button
  aria-label="bitcoin tag"
  aria-pressed="false"
  class="cursor-pointer inline-flex items-center justify-center bg-transparent focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 min-w-9 group relative h-8 gap-1 rounded-lg px-3 backdrop-blur-lg border-0 text-sm font-bold leading-5 text-white transition-all duration-200 hover:bg-transparent hover:text-white data-[state=on]:border data-[state=on]:border-solid data-[state=on]:bg-transparent data-[state=on]:text-white"
  data-state="off"
  type="button"
>
  <span
    class="flex items-center gap-1.5 text-sm leading-5"
  >
    <span
      class="font-bold"
    >
      bitcoin
    </span>
  </span>
  <div
    aria-hidden="true"
    class="pointer-events-none absolute inset-0 rounded-lg opacity-0 transition-opacity group-hover:opacity-100"
    style="box-shadow: inset 0px 0px 8px 0px #FF9900;"
  />
</button>
`;
